### PR TITLE
GPU-1711: CMS: Page Expire Date

### PR DESCRIPTION
**Pull request by guggi**
Epic: https://jira-gentics.atlassian.net/browse/GPU-1711

### CODE REVIEW

File: cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishLogEntry.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	/**
+	 * Loads a publish log entry by a specified field and value.
+	 *
+	 * @param fieldName the name of the field
+	 * @param value the value of the field
+	 * @return an Optional containing the publish log entry if found, otherwise empty
+	 * @throws NodeException if an error occurs during the load operation
+	 */
+	public Optional<PublishLogEntry> loadByField(String fieldName, int value) throws NodeException {
+		var query = String.format("SELECT * FROM publish_protocol WHERE %s = %s ORDER BY id DESC", fieldName, value);
```
Comment (guggi) [inline cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishLogEntry.java:109]: add type to differentiate between form and page

File: cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishLogEntry.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+				LocalDateTime dateTime = timestamp.toLocalDateTime();
+
+				entries.add(new PublishLogEntry(id, objId, type, state, user, dateTime));
+			}
+			return entries;
+		});
+	}
+
+
+	public int getId() {
```
Comment (guggi) [inline cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishLogEntry.java:150]: add javadoc

File: cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishProtocolService.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	 *
+	 * @param <T>    the type of the publishable node object
+	 * @param object the publishable node object
+	 * @param status the status of the publish operation
+	 * @param user   the user performing the operation
+	 * @throws NodeException if an error occurs during logging
+	 */
+	public <T extends PublishableNodeObject> void logPublishState(T object, int status, int user)
+			throws NodeException {
+		var publishLogEntry = new PublishLogEntry(object.getId(), getType(object.getTType()), status,
```
Comment (guggi) [inline cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishProtocolService.java:32]: maybe use ttype directly

File: cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishProtocolService.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	/**
+	 * Retrieves a publish log entry by its object ID.
+	 *
+	 * @param objId the object ID of the publish log entry
+	 * @return the publish log entry
+	 * @throws NodeException           if an error occurs during retrieval
+	 * @throws EntityNotFoundException if the entry is not found
+	 */
+	public PublishLogEntry getPublishLogEntryByObjectId(int objId) throws NodeException {
+		return new PublishLogEntry().loadByField("obj_id", objId).orElseThrow(
```
Comment (guggi) [inline cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishProtocolService.java:50]: add type to retrieve correct object

File: cms-core/src/main/java/com/gentics/contentnode/rest/resource/impl/PublishProtocolResourceImpl.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+			publishLogEntry.getType(),
+			publishLogEntry.getState() == 1 ? "ONLINE" : "OFFLINE",
+			publishLogEntry.getUser(),
+			publishLogEntry.getDate().toString()
+	);
+
+	@Override
+	@GET
+	@Path("/{objId}")
+	public PublishLogDto get(@PathParam("objId") Integer objId) throws NodeException {
```
Comment (guggi) [inline cms-core/src/main/java/com/gentics/contentnode/rest/resource/impl/PublishProtocolResourceImpl.java:53]: add type here as well

File: cms-core/src/main/resources/changelog/system.CHANGELOG [hunk 6]
```diff
@@ -297,3 +297,5 @@
 2023-03-21  CH-1689 NOP SQL DROP TABLE IF EXISTS bundleimportconflict;
 2023-03-21  CH-1690 NOP SQL DROP TABLE IF EXISTS bundleimportobject;
 2024-03-29  CH-1717 SP  SQL DELETE FROM invokerqueue WHERE type NOT IN ('sleep', 'nodecopy', 'error');
+2024-09-02  CH-1724 GO  SQL CREATE TABLE publish_protocol (id INT NOT NULL PRIMARY KEY AUTO_INCREMENT, obj_id INT NOT NULL, type VARCHAR(50) NOT NULL, state TINYINT NOT NULL, user INT NOT NULL, date DATETIME NOT NULL DEFAULT CURRENT_DATE());
+2024-09-02  CH-1725 GO  SQL CREATE INDEX publish_protocol_obj_id_index ON publish_protocol (obj_id);
```
Comment (guggi) [inline cms-core/src/main/resources/changelog/system.CHANGELOG:301]: add type to index

File: cms-core/src/main/resources/contentnode_de_DE.properties [hunk 7]
```diff
@@ -9,6 +9,7 @@ actions=Aktionen
 log=Protokoll
 rest.general.error=Es ist ein Fehler beim Zugriff auf das Backend System aufgetreten.
 rest.invalid.json.error=Das Format des Request Bodies war inkorrekt.
+rest.invalid.request=Ung�ltige Anfrage
```
Comment (guggi) [inline cms-core/src/main/resources/contentnode_de_DE.properties:12]: double check if unicode escaping is needed

File: cms-oss-changelog/src/changelog/entries/2024/09/7948.GPU-1711.enhancement [hunk 8]
```diff
@@ -0,0 +1,5 @@
+Page Expire Date: After unpublishing/taking a page in the CMS offline, the information of when it went offline is now available to the user.
+Two new endpoints have been introduced to obtain this information
+
+* GET /publish/state/{objId}
+* GET publish/state/
```
Comment (guggi) [inline cms-oss-changelog/src/changelog/entries/2024/09/7948.GPU-1711.enhancement:5]: slash is missing

File: cms-oss-changelog/src/changelog/entries/2024/09/7948.GPU-1711.enhancement [hunk 9]
```diff
@@ -0,0 +1,5 @@
+Page Expire Date: After unpublishing/taking a page in the CMS offline, the information of when it went offline is now available to the user.
+Two new endpoints have been introduced to obtain this information
+
+* GET /publish/state/{objId}
```
Comment (guggi) [inline cms-oss-changelog/src/changelog/entries/2024/09/7948.GPU-1711.enhancement:4]: add:   
*@GET /publish/state/{objId}@

File: cms-core/src/main/java/com/gentics/contentnode/factory/object/ExtensiblePublishableObject.java [hunk 10]
```diff
@@ -14,6 +15,8 @@ public interface ExtensiblePublishableObject<T extends PublishableNodeObject> ex
 	@Override
 	List<? extends ExtensiblePublishableObjectService<T>> getServices();
 
+	PublishProtocolService publishProtocolService = new PublishProtocolService();
```
Comment (guggi) [inline cms-core/src/main/java/com/gentics/contentnode/factory/object/ExtensiblePublishableObject.java:18]: maybe make the method static and avoid instantiating this potentially often

File: cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishState.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package com.gentics.contentnode.publish.protocol;
+
+public enum PublishState {
```
[the remark this answers is not in the document]
Comment (guggi) [reply cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishState.java:3]: d487044f

File: cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishProtocolService.java (hunk 4 above)
Comment (guggi) [reply cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishProtocolService.java]: d487044f

File: cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishLogEntry.java (hunk 2 above)
Comment (guggi) [reply cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishLogEntry.java:154]: d487044f

File: cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishLogEntry.java (hunk 1 above)
Comment (guggi) [reply cms-core/src/main/java/com/gentics/contentnode/publish/protocol/PublishLogEntry.java]: d487044f

File: cms-core/src/main/java/com/gentics/contentnode/factory/object/ExtensiblePublishableObject.java (hunk 10 above)
Comment (guggi) [reply cms-core/src/main/java/com/gentics/contentnode/factory/object/ExtensiblePublishableObject.java]: d487044f

File: cms-oss-changelog/src/changelog/entries/2024/09/7948.GPU-1711.enhancement (hunk 8 above)
Comment (guggi) [reply cms-oss-changelog/src/changelog/entries/2024/09/7948.GPU-1711.enhancement]: d487044f

File: cms-oss-changelog/src/changelog/entries/2024/09/7948.GPU-1711.enhancement (hunk 9 above)
Comment (guggi) [reply cms-oss-changelog/src/changelog/entries/2024/09/7948.GPU-1711.enhancement]: d487044f

File: cms-core/src/main/resources/contentnode_de_DE.properties (hunk 7 above)
Comment (guggi) [reply cms-core/src/main/resources/contentnode_de_DE.properties:12]: i had no issues using those characters without escapint

File: cms-core/src/main/resources/changelog/system.CHANGELOG (hunk 6 above)
Comment (guggi) [reply cms-core/src/main/resources/changelog/system.CHANGELOG]: d487044f

File: cms-core/src/main/java/com/gentics/contentnode/rest/resource/impl/PublishProtocolResourceImpl.java (hunk 5 above)
Comment (guggi) [reply cms-core/src/main/java/com/gentics/contentnode/rest/resource/impl/PublishProtocolResourceImpl.java]: d487044f

File: cms-core/src/main/java/com/gentics/contentnode/factory/object/FormFactory.java [hunk 12]
```diff
@@ -329,13 +339,47 @@ public ContentNodeDate getPDate() {
 			return pDate;
 		}
 
+		@Override
+		public ContentNodeDate getUnpublishedDate() {
+			try {
```
Comment (guggi) [inline cms-core/src/main/java/com/gentics/contentnode/factory/object/FormFactory.java:344]: try remove this and align to existing approach

File: cms-core/src/main/java/com/gentics/contentnode/factory/object/PageFactory.java [hunk 13]
```diff
@@ -2599,6 +2635,9 @@ public void takeOffline(int timestamp) throws ReadOnlyException, NodeException {
 				// we need to sent the NOTIFY event for the page in order to allow indexing (for feature ELASTICSEARCH)
 				t.addTransactional(new TransactionalTriggerEvent(Page.class, getId(), INDEXED_STATUS_ATTRIBUTES, Events.NOTIFY));
 			}
+			Transaction t = TransactionManager.getCurrentTransaction();
```
Comment (guggi) [inline cms-core/src/main/java/com/gentics/contentnode/factory/object/PageFactory.java:2638]: align as well

File: cms-restapi/src/main/java/com/gentics/contentnode/rest/model/ContentNodeItem.java [hunk 14]
```diff
@@ -289,4 +295,23 @@ public Integer getUnpublishedDate() {
...
+	}
+
+	/**
+	 * Sets the unpublisher.
+	 *
+	 * @param unpublisher the unpublisher to set
+	 */
+	public void setUnpublisher(User unpublisher) {
+		this.unpublisher = unpublisher;
+	}
```
Comment (guggi) [inline cms-restapi/src/main/java/com/gentics/contentnode/rest/model/ContentNodeItem.java:315]: move to page and form or common base class

**Review by deckdom — APPROVED** [review]

**Review by deckdom — APPROVED** [review]

**Pull request merged**